.ci/new_ci.jenkinsfile

```diff
@@ -0,0 +1,435 @@
+        sh 'docker login --username $USERNAME --password $PASSWORD ' + "${registry}"
+    }
+
+    sh "docker build -t ${image["name"]} ${buildArgs} ${target} ."
+}
+
+def buildModernImages(List<Map> images, String registry, String registryCredentialsID) {
+    // Define helper function to build the compiler
+    Closure imageBuild = { Map image ->
+        return {
+            node('Linux') {
+                String buildArgsStr = image["buildArgs"].collect({ k, v -> "--build-arg ${k }=${v }" }).join(' ') as String
+                String target = image["target"] != null ? "--target ${image["target"]}" : ""
+
+                environment {
+                    DOCKER_BUILDKIT = 0
+                    COMPOSE_DOCKER_CLI_BUILD = 0
+                }
+
+                stage('Modern - Checkout sources and install deps') {
+                    checkout scm
+                    // sh 'pip install pytest'
+                    // TODO: Pass expected in a different way
+                    // writeFile(file: 'modern/.env', text: testArguments.collect({ k, v -> "${k }=${v }"}).join('\n') as String)
+                    // sh 'cat modern/.env'
+                }
+
+                stage("Modern - Build image ${image["name"]}") {
+                    dir(image["path"]) {
+                        buildModernImage(image, buildArgsStr, target, registry, registryCredentialsID)
+                    }
+                }
+
+                stage("Modern - Upload image ${image["name"]} to Artifactory") {
+                    uploadImage(image, registry, registryCredentialsID)
+                }
+            }
+        }
+    }
+
+    // Build images in parallel
+    // base, builder, normal, jenkins
+    List<Map> baseImages = images.findAll({ Map image -> image["name"].contains("base") })
+    List<Map> builderImages = images.findAll({ Map image -> image["name"].contains("builder") })
+    List<Map> normalImages = images.findAll({ Map image -> !image["name"].contains("base") && !image["name"].contains("builder") && !image["name"].contains("jenkins") })
+    List<Map> jenkinsImages = images.findAll({ Map image -> image["name"].contains("jenkins") })
+    buildParallel("Modern base image", baseImages, imageBuild)
+    buildParallel("Modern builder image", builderImages, imageBuild)
+    buildParallel("Modern normal image", normalImages, imageBuild)
+    buildParallel("Modern image jenkins", jenkinsImages, imageBuild)
+}
+
+def promoteImages(List<Map> images, String fromRegistryName, String fromRegistryCredentialsId,
+                  String toRegistryName, String toRegistryCredentialsId) {
+    withCredentials([usernamePassword(credentialsId: fromRegistryCredentialsId, usernameVariable: 'USERNAME', passwordVariable: 'PASSWORD')]) {
+        sh 'docker login --username $USERNAME --password $PASSWORD ' + "${fromRegistryName}"
+    }
+
+    images.each { Map image ->
```

Comment:
In the future we need to add a retry here. Networking errors are not uncommon, and needing to build all again due a single failure is a hell.

Reply:
totally, I will add a TODO